### PR TITLE
Load schemas on demand

### PR DESCRIPTION
**Pull request by andreineculau**
I wanted to address only #23 at first, but when cleaning up, I involuntarily ended up tackling the todos in https://github.com/for-GET/jesse/issues/13#issuecomment-190431727

In short, this messy PR (due to the commit history and lack of new tests)
- loads schemas on demand (the initial validation schema needs an `id` though) within the file/http/https schemes
- uses fully-qualified keys for the schemas, meaning these are equivalent "./test1.json", "test1.json", "foo/../test1.json", etc
- schemas are stored internally as
  - `SourceKey` e.g. file://path/file.json
  - `Key` the schema's `id`
  - Modified time
  - Schema
- the CLI version is now working closer to the expectations (loading schemas from the filesystem on demand)

### CODE REVIEW

File: src/jesse_state.erl [hunk 1]
```diff
@@ -303,6 +358,7 @@ find_schema(#state{schema_loader_fun=LoaderFun}, SchemaURI) ->
             ?not_found
         end
   catch
-    _:_ ->
+    _C:_E ->
+      %% io:format("load_schema failure: ~p\n", [{_C, _E, erlang:get_stacktrace()}]),
```
Comment (gadgetci) [inline src/jesse_state.erl]: According to **Elvis**:
> Line 362 is too long:       %% io:format("load_schema failure: ~p\n", [{_C, _E, erlang:get_stacktrace()}]),.

File: src/jesse_database.erl [hunk 2]
```diff
@@ -30,171 +30,282 @@
...
+  SchemaInfos = [{Key, 0, Schema}],
+  ValidationFun = fun jesse_lib:is_json_object/1,
+  store_schemas(SchemaInfos, ValidationFun);
+add_uri("https://" ++ _ = Key) ->
+  {ok, Response} = httpc:request(get, {Key, []}, [], [{body_format, binary}]),
+  {{_Line, 200, _}, _Headers, Body} = Response,
+  Schema = jsx:decode(Body),
+  SchemaInfos = [{Key, 0, Schema}],
+  ValidationFun = fun jesse_lib:is_json_object/1,
+  store_schemas(SchemaInfos, ValidationFun);
```
Comment (seriyps) [inline src/jesse_database.erl]: http and https clauses are like... copy-pasted?

File: src/jesse_database.erl [hunk 3]
```diff
@@ -30,171 +30,282 @@
...
+  {ok, #file_info{mtime = Mtime}} = file:read_file_info(File),
+  Schema = jsx:decode(Body),
+  SchemaInfos = [{Key, Mtime, Schema}],
+  ValidationFun = fun jesse_lib:is_json_object/1,
+  store_schemas(SchemaInfos, ValidationFun);
+add_uri("http://" ++ _ = Key) ->
+  {ok, Response} = httpc:request(get, {Key, []}, [], [{body_format, binary}]),
+  {{_Line, 200, _}, _Headers, Body} = Response,
+  Schema = jsx:decode(Body),
+  SchemaInfos = [{Key, 0, Schema}],
```
Comment (seriyps) [inline src/jesse_database.erl]: We may TRY to extract mtime from 'Last-Modified' HTTP header

**Comment by andreineculau** [comment]
@seriyps thanks for the comments. If there's more, keep it coming.

File: src/jesse.erl [hunk 4]
```diff
@@ -84,25 +83,21 @@ del_schema(Key) ->
...
+%% Equivalent to `load_schemas(Path, ParserFun, ValidationFun)'
+%% where `ValidationFun' is `fun jesse_json:is_json_object/1'.
 -spec load_schemas( Path      :: string()
                   , ParserFun :: fun((binary()) -> json_term())
                   ) -> jesse_database:update_result().
 load_schemas(Path, ParserFun) ->
   load_schemas( Path
               , ParserFun
               , fun jesse_lib:is_json_object/1
-              , fun jesse_lib:get_schema_id/1
```
Comment (lazedo) [inline src/jesse.erl]: @andreineculau why are you changing arity ? we currently use arity/4 with our get_schema_id fun.
it would be great if you could maintain compatibility.

File: src/jesse_lib.erl [hunk 5]
```diff
@@ -25,7 +25,6 @@
 
 %% API
 -export([ empty_if_not_found/1
-        , get_schema_id/1
```
Comment (lazedo) [inline src/jesse_lib.erl]: in some use cases we default from our fun to get_schema_id/1. see comment above about load_schemas arity

File: src/jesse_state.erl [hunk 6]
```diff
@@ -61,7 +62,7 @@
                                      ) -> list() | no_return()
                                             )
          , default_schema_ver :: binary()
-         , schema_loader_fun  :: fun(( binary()
+         , schema_loader_fun  :: fun(( string()
```
Comment (lazedo) [inline src/jesse_state.erl]: why this change here ? can we maintain compatibility ?
there are already a lot of match clauses depending on binary

File: src/jesse_state.erl [hunk 7]
```diff
@@ -36,9 +36,10 @@
         , set_allowed_errors/2
         , set_current_schema/2
         , set_error_list/2
-        , find_schema/2
```
Comment (lazedo) [inline src/jesse_state.erl]: why do you remove this function ? library users may be using it, and it seems you only rename it to  `load_schema`

File: src/jesse.erl (hunk 4 above)
Comment (andreineculau) [inline src/jesse.erl]: @lazedo how does your `get_schema_id` look like?

It's been a while since I filed this PR, but the idea is 1) either your schema file has an `id`, 2) either it doesn't and then we default the id to the path of the schema file itself. FYI the schema can be referenced from another schema both by it's id and it's filesystem location.

Let me know what's your usecase, and we can figure out smth.

File: src/jesse_state.erl (hunk 6 above)
Comment (andreineculau) [inline src/jesse_state.erl]: correct me if I'm wrong, but this is just a spec correction, because even before this PR, the `Key` (argument to schema_loader_fun) was a string, since that is what `get_schema_id/1` was returning.

You can even see that before this PR, we were converting binary to string before calling `schema_loader_fun` https://github.com/for-GET/jesse/pull/24/files#diff-70cc40937373cde9cea9266b655884c0L292

File: src/jesse_state.erl (hunk 7 above)
Comment (andreineculau) [inline src/jesse_state.erl]: The jesse API is the one defined in `jesse.erl`.

If you have a usecase, let me know and we can look into this as well, otherwise I can't work with "what ifs", and pollute the codebase with aliases and/or backward-compatibility wrappers because maybe somewhere someone somehow would depend on x and they cannot be bothered to run `sed "s/x/y/g"`.

**Comment by andreineculau** [comment]
@lazedo thanks a bunch for having a look and providing feedback!

**Comment by andreineculau** [comment]
**TODO: FIX ISSUE RE resolve_ref**

https://github.com/for-GET/jesse/commit/a1ab58c42b98152d7b61730d0a456c1bb5c1a29a#commitcomment-18957182

**Comment by andreineculau** [comment]
ping @lazedo 
I intend to close up soon (before xmas). Would you be so kind to give some info regarding get_schema_id ? I don't know what was @Cy6erBr4in 's rationale for the `MakeKeyFun` functionality i.e. override custom schema id, so if you have a strong case, I can spend some time keeping it. At face value, I cannot imagine the reason for it *given* that one can load schemas on demand, which this PR solves.

I'll have a look at https://github.com/for-GET/jesse/commit/a1ab58c42b98152d7b61730d0a456c1bb5c1a29a#commitcomment-18957182 meanwhile

**Comment by andreineculau** [comment]
one last ping @lazedo

**Comment by lazedo** [comment]
@andreineculau hi, sorry about bad response time. i'll take a look until tuesday 20, if that's ok with you. also , i believe there are still some issues with draft4 validator for which i have a working solution. i'll review it with latest master and let you know. if you wish, you may take a look at https://github.com/lazedo/jesse/tree/1.4.3
Best

**Comment by andreineculau** [comment]
@lazedo great! and no worries (sorry if I felt pushy). 25th is my deadline

**Comment by andreineculau** [comment]
@lazedo time is way up, so i'm going to merge this, but your feedback is very much appreciated anytime
